File: python/linked-list.py
# slow but sturdy linked list

class MyLinkedList:

    def __init__(self):
        self.head = '0'
        self.length = 0
        self.pointerCounter = 10000

    def __getitem__(self, key):
        return getattr(self, key)

    def getPointer(self):
        self.pointerCounter += 1
        return str(self.pointerCounter)

    def get(self, index: int) -> int:
        # find pointer at index
        if ((index >= self.length) | (index < 0)): 
            return -1
        lookAhead = self.head
        for i in range(0, index):
            lookAhead = self[lookAhead]['nextPtr'] 

        return self[lookAhead]['value']
            
    def addAtHead(self, val: int) -> None:
        # create new pointer and value/next pair
        pair = {'value': val, 'nextPtr': self.head}
        newPoint = self.getPointer()
        setattr(self, newPoint, pair)
        
        # update head
        self.head = newPoint
        
        # update length
        self.length += 1

    def addAtTail(self, val: int) -> None:
        # create new pointer and value/next pair
        pair = { 'value': val, 'nextPtr': '0' }
        newPoint = self.getPointer()
        setattr(self, newPoint, pair)
        
        # update tail
        tracker = [self.head, '-1']
        while(tracker[0] != '0'):
            tracker[1] = tracker[0]
            tracker[0] = self[tracker[0]]['nextPtr']
        if (tracker[1] != '-1'):
            self[tracker[1]]['nextPtr'] = newPoint
        else:
            self.head = newPoint

        # update length
        self.length += 1
        return

    def addAtIndex(self, index: int, val: int) -> None:
        # handle too high index
        if (index > self.length):
            return
    
        # handle index equals length
        if (index == self.length):
            self.addAtTail(val)
            return

        # find pointer for indexed item
        tracker = [self.head, '-1']
        for i in range(0, index):
            tracker[1] = tracker[0]
            tracker[0] = self[tracker[0]]['nextPtr']
        
        # create new pointer and value/next pair
        pair = { 'value': val, 'nextPtr': tracker[0] }
        newPoint = self.getPointer()
        setattr(self, newPoint, pair)

        # update the item before the new one        
        if (tracker[1] != '-1'):
            self[tracker[1]]['nextPtr'] = newPoint
        else:
            self.head = newPoint
            
        # update length
        self.length += 1
        return

    def deleteAtIndex(self, index: int) -> None:
        if (index >= self.length):
            return
        
        # find pointers before and at index
        tracker = [self.head, '-1']
        for i in range(0, index):
            tracker[1] = tracker[0]
            tracker[0] = self[tracker[0]]['nextPtr']

        # update the item before the deleted one
        if (tracker[1] != '-1'):
            self[tracker[1]]['nextPtr'] = self[tracker[0]]['nextPtr']
        else:
            self.head = self[tracker[0]]['nextPtr']
        
        # remove item from object
        delattr(self, tracker[0])
        
        # update length
        self.length -= 1
        
        return

obj = MyLinkedList()
obj.addAtHead(300)
obj.addAtHead(200)
obj.addAtHead(100)
print(vars(obj))
print('-----------')
print(obj.get(0))
print(obj.get(1))
print(obj.get(2))
print(obj.get(3))
print(obj.get(-1))
print('-----------')
obj.addAtTail(400)
print(vars(obj))
print('-----------')
obj.addAtIndex(2, 250)
print(vars(obj))
print('-----------')
obj.deleteAtIndex(3)
obj.deleteAtIndex(0)
print(vars(obj))

File: archive/linked-list.js
// slow but sturdy

class MyLinkedList {
  constructor() {
    this.head = 0; // zero when nothing in list
    this.length = 0;
    this.pointerCounter = 10000;
  }

  nextPointer(pointer) {
    return [this[pointer].next, pointer];
  }

  nextPointer() {
    this.pointerCounter++;
    return this.pointerCounter;
  }
}

/**
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function (index) {
  // find pointer at index
  if (index >= this.length || index < 0 || !Number.isInteger(index)) return -1;

  let lookAhead = this.head;
  for (let i = 0; i < index; i++) {
    lookAhead = this[lookAhead].next;
  }
  return this[lookAhead].value;
};

/**
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function (val) {
  // create new pointer and value/next pair
  let newPoint = this.nextPointer();
  let newPair = {
    value: val,
    next: this.head,
  };
  this[newPoint] = newPair;
  // update head
  this.head = newPoint;
  // update length
  this.length++;
};

/**
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function (val) {
  // create new pointer and value/next pair
  let newPoint = this.nextPointer();
  let newPair = {
    value: val,
    next: 0, // because this pair is the new tail
  };
  this[newPoint] = newPair;

  // update tail
  let lookAhead = this.head;
  let lastPointer = -1;
  while (lookAhead != 0) {
    lastPointer = lookAhead;
    lookAhead = this[lookAhead].next;
  }
  if (lastPointer != -1) {
    this[lastPointer].next = newPoint;
  } else {
    this.head = newPoint;
  }

  // update length
  this.length++;
};

/**
 * @param {number} index
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function (index, val) {
  // find pointers before and at index
  if (index > this.length) {
    console.log("addAt not valid", index);
    return;
  }

  if (index == this.length) {
    console.log("equal to length", index, val);
    this.addAtTail(val);
    return;
  }

  let lookAhead = this.head;
  let lastPointer = -1;
  for (let i = 0; i < index; i++) {
    lastPointer = lookAhead;
    lookAhead = this[lookAhead].next;
  }

  // create new pointer and value/next pair
  let newPoint = this.nextPointer();
  let newPair = {
    value: val,
    next: lookAhead,
  };
  this[newPoint] = newPair;

  // update the item before the new one
  if (lastPointer != -1) {
    this[lastPointer].next = newPoint;
  } else {
    this.head = newPoint;
  }

  // update length
  this.length++;
};

/**
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function (index) {
  // find pointers before and at index
  if (index >= this.length) return;

  let lookAhead = this.head;
  let lastPointer = -1;
  for (let i = 0; i < index; i++) {
    lastPointer = lookAhead;
    lookAhead = this[lookAhead].next;
  }

  // update the item before the deleted one
  if (lastPointer != -1) {
    this[lastPointer].next = this[lookAhead].next;
  } else {
    this.head = this[lookAhead].next;
  }

  // remove item from this
  delete this[lookAhead];

  // update length
  this.length--;
};

obj = new MyLinkedList();
obj.addAtHead(300);
obj.addAtHead(200);
obj.addAtHead(100);
console.log(obj);
console.log("-----------");
console.log(obj.get(0));
console.log(obj.get(1));
console.log(obj.get(2));
console.log(obj.get(3));
console.log(obj.get(-1));
console.log("-----------");
obj.addAtTail(400);
console.log(obj);
console.log("-----------");
obj.addAtIndex(2, 250);
console.log(obj);
console.log("-----------");
obj.deleteAtIndex(3);
obj.deleteAtIndex(0);
console.log(obj);
